pkg/dtest/testprocess/tp.go

```diff
@@ -72,50 +78,60 @@ var singleton = &sub{false, make(map[string]func())}
 //         err := fooCmd.Run()
 //         ...
 //     }
+//
+// It is permissible, but not required, to call flag.Parse before
+// calling testprocess.Dispatch.  If flag.Parse has not been called,
+// then testprocess.Dispatch will call it.
 func Dispatch() {
-	// if this ever causes problems, we can switch back to getting
-	// the name from the TESTPROCESS_NAME environment variable
-	name := flag.String("testprocess.name", "", "")
-	flag.Parse()
+	dispatched = true
 
+	if !flag.Parsed() {
+		flag.Parse()
+	}
 	if *name == "" {
-		if singleton.sudo && os.Geteuid() != 0 {
-			dtest.Sudo()
-		}
-	} else {
-		log.Printf("TESTPROCESS %s PID: %d", *name, os.Getpid())
+		return
+	}
+
+	log.Printf("TESTPROCESS %s PID: %d", *name, os.Getpid())
 
-		defer func() {
-			if r := recover(); r != nil {
-				stack := string(debug.Stack())
-				log.Printf("TESTPROCESS %s PANICKED: %v\n%s", *name, r, stack)
-				os.Exit(1)
-			}
-		}()
+	defer func() {
+		if r := recover(); r != nil {
+			stack := string(debug.Stack())
+			log.Printf("TESTPROCESS %s PANICKED: %v\n%s", *name, r, stack)
+			os.Exit(1)
+		}
+	}()
 
-		singleton.functions[*name]()
+	functions[*name]()
 
-		log.Printf("TESTPROCESS %s NORMAL EXIT", *name)
-		os.Exit(0)
-	}
+	log.Printf("TESTPROCESS %s NORMAL EXIT", *name)
+	os.Exit(0)
 }
 
 // Make returns an *exec.Cmd that will execute the supplied function
 // in a subprocess. For this to work, testprocess.Dispatch must be
-// invoked by the TestMain of any test suite using this, and the
-// call to Make must be from a global variable initializer, e.g.:
+// invoked by the TestMain of any test suite using this, and the call
+// to Make must be *before* the call to testprocess.Dispatch; possibly
+// from a global variable initializer, e.g.:
 //
 //     var myCmd = testprocess.Make(func() { doSomething(); })
 //
 func Make(f func()) *exec.Cmd {
-	return singleton._make(false, f)
+	if dispatched {
+		// panic because it's a bug in the code, and a stack
+		// trace is useful.
+		panic("testprocess: testprocess.Make called after testprocess.Dispatch")
+	}
```

Comment:
Bleh. If I want to recreate the same Cmd to call it a second time, this means I need to come up with a different way to do that.

Reply:
why can't you just create both instances at the beginning?
```go
var (
	cmd1 = testprocess.Make(fn)
	cmd2 = testprocess.Make(fn)
)
```

I viewed that change as just enforcing the already-documented contract, to make bugs fail and be easier to detect.

Reply:
I have some code that polls for a state change, e.g., https://github.com/datawire/teleproxy/pull/167/files#diff-2d1c0e57b402f3fc3e3695eac2437280R210-R220. That being said, because I have to build the `edgectl` binary for the test, I suppose I can skip using this functionality entirely and just call the binary for this case.